### ISSUE TITLE
DataStore can't query a model with contains predicate

### ISSUE THREAD

**Issue by DevTGhosh** [aws-amplify/amplify-js#8262]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

React

### Amplify APIs

DataStore

### Environment information

<details>

```
# Put output below this line
 System:
    OS: Linux 4.19 Ubuntu 20.04.1 LTS (Focal Fossa)
    CPU: (8) x64 Intel(R) Core(TM) i5-9300H CPU @ 2.40GHz
    Memory: 6.50 GB / 12.36 GB
    Container: Yes
    Shell: 5.8 - /usr/bin/zsh
  Binaries:
    Node: 12.18.4 - ~/.nvm/versions/node/v12.18.4/bin/node
    npm: 6.14.9 - ~/.nvm/versions/node/v12.18.4/bin/npm
  npmPackages:
    @aws-amplify/api: ^3.2.17 => 3.2.29 
    @aws-amplify/auth: ^3.4.17 => 3.4.29 
    @aws-amplify/core: ^3.8.9 => 3.8.21 
    @aws-amplify/datastore: ^2.9.14 => 2.9.15 
    @aws-amplify/storage: ^3.3.17 => 3.3.29 
    @hookform/devtools: ^2.2.1 => 2.2.1 
    @hookform/error-message: 0.0.5 => 0.0.5 
    @material-ui/core: ^4.11.1 => 4.11.4 
    @material-ui/icons: ^4.9.1 => 4.11.2 
    @material-ui/lab: ^4.0.0-alpha.57 => 4.0.0-alpha.58 
    @react-pdf/renderer: ^1.6.13 => 1.6.16 
    @reduxjs/toolkit: ^1.5.0 => 1.5.1 
    @testing-library/jest-dom: ^5.11.6 => 5.12.0 
    @testing-library/react: ^11.2.2 => 11.2.6 
    @testing-library/user-event: ^12.6.0 => 12.8.3 
    @typescript-eslint/eslint-plugin: ^4.18.0 => 4.22.1 
    @typescript-eslint/parser: ^4.18.0 => 4.22.1 
    aws-amplify: ^3.3.25 => 3.3.26 
    babel-eslint: ^10.1.0 => 10.1.0 
    clsx: ^1.1.1 => 1.1.1 
    dayjs: ^1.10.3 => 1.10.4 
    eslint: ^7.22.0 => 7.25.0 
    eslint-config-react-app: ^6.0.0 => 6.0.0 
    eslint-plugin-flowtype: ^5.4.0 => 5.7.2 
    eslint-plugin-import: ^2.22.1 => 2.22.1 
    eslint-plugin-jsx-a11y: ^6.4.1 => 6.4.1 
    eslint-plugin-react: ^7.22.0 => 7.23.2 
    eslint-plugin-react-hooks: ^4.2.0 => 4.2.0 
    fontsource-roboto: ^3.0.3 => 3.1.5 
    history: ^5.0.0 => 5.0.0 
    husky: ^4.3.8 => 4.3.8 
    i18next: ^19.8.4 => 19.9.2 
    i18next-browser-languagedetector: ^6.0.1 => 6.1.0 
    i18next-http-backend: ^1.1.0 => 1.2.2 
    lint-staged: ^10.5.4 => 10.5.4 
    lodash: ^4.17.20 => 4.17.21 
    material-ui-dropzone: ^3.5.0 => 3.5.0 
    memo-parser:  0.2.1 
    prettier: ^2.2.1 => 2.2.1 
    react: ^17.0.1 => 17.0.2 
    react-data-grid: 7.0.0-alpha.29 => 7.0.0-alpha.29 
    react-data-grid-addons: ^7.0.0-alpha.24 => 7.0.0-alpha.24 
    react-dom: ^17.0.1 => 17.0.2 
    react-hook-form: ^6.12.1 => 6.15.7 
    react-i18next: ^11.8.4 => 11.8.15 
    react-infinite-scroll-component: ^5.1.0 => 5.1.0 
    react-redux: ^7.2.2 => 7.2.4 
    react-router: ^6.0.0-beta.0 => 6.0.0-beta.0 
    react-router-dom: ^6.0.0-beta.0 => 6.0.0-beta.0 
    react-scripts: 4.0.1 => 4.0.1 
    react-toastify: ^7.0.3 => 7.0.4 
    react-virtualized: ^9.22.2 => 9.22.3 
    redux: ^4.0.5 => 4.1.0 (3.7.2)
    web-vitals: ^0.2.4 => 0.2.4 
    workbox-background-sync: ^5.1.4 => 5.1.4 
    workbox-broadcast-update: ^5.1.4 => 5.1.4 
    workbox-cacheable-response: ^5.1.4 => 5.1.4 
    workbox-core: ^5.1.4 => 5.1.4 
    workbox-expiration: ^5.1.4 => 5.1.4 
    workbox-google-analytics: ^5.1.4 => 5.1.4 
    workbox-navigation-preload: ^5.1.4 => 5.1.4 
    workbox-precaching: ^5.1.4 => 5.1.4 
    workbox-range-requests: ^5.1.4 => 5.1.4 
    workbox-routing: ^5.1.4 => 5.1.4 
    workbox-strategies: ^5.1.4 => 5.1.4 
    workbox-streams: ^5.1.4 => 5.1.4 
  npmGlobalPackages:
    @aws-amplify/cli: 4.50.2
    expo-cli: 4.0.13
    migrate-mongoose: 4.0.0
    nodemon: 2.0.5
    npm: 6.14.9
    prettier: 2.1.2

```

</details>

### Describe the bug

Datastore is failing silently when trying to query a model using the contains predicates. It doesn't return a result and rest of the function doesn't fire. 

This is similar to the issue #7996 . I have commented over there but it hasn't been looked over in some time & it's a major blocker for my company. If anyone can take  a look at it would be really helpful. 

### Expected behavior

Searches for datastore items that contains the value in an field array & returns the array of results. 

### Reproduction steps

1. Create a model Project with a field(freelancers) with array of ids 
2. Use datastore to query the model Project where you check if  field freelancers contains a particular id

### Code Snippet

My Model
```graphql
type Projects
  @model(timestamps: { createdAt: "createdOn", updatedAt: "updatedOn" })
 {
  id: ID!
  name: String!
  tenantId: ID!
  freelancers: [ID]
  childFreelancers: [ID]
  parentProjectId: ID
  parentClientId: ID
  activeFlag: Boolean!
  createdOn: AWSDateTime
  updatedOn: AWSDateTime
}
```

My code 
```js
console.log('auth.userInfo.tenantId: ', auth.userInfo.tenantId)
console.log('check')
const projectData = await DataStore.query(Projects, c =>
        c.freelancers('contains', auth.userInfo.tenantId).activeFlag('eq', true)
)

//The below 2 console.log don't fire when contains is used
console.log('check2')
console.log('projectData: ', projectData)
```

### Log output

<details>

```
[DEBUG] 46:16.150 Hub - Dispatching to datastore with  
{event: "ready", data: undefined}

// Console.log to check
projects.js:35 auth.userInfo.tenantId:  02487b56-235d-4541
projects.js:37 check

ConsoleLogger.ts:99 [DEBUG] 46:16.155 DataStore - params ready 
{predicate: {…}, pagination: {…}, modelConstructor: ƒ}

modelConstructor: ƒ Model(init)
copyOf: ƒ (source, fn)
fromJSON: ƒ (json)
Symbol(immer-draftable): true
arguments: (...)
caller: (...)
length: 1
name: "Projects"
prototype: {constructor: ƒ}
__proto__: ƒ ()

[[FunctionLocation]]: datastore.ts:352
[[Scopes]]: Scopes[3]
pagination: {sort: undefined}
predicate:
predicates: Array(2)
0: {field: "freelancers", operator: "contains", operand: "02487b56-235d-4541"}
1: {field: "activeFlag", operator: "eq", operand: true}
length: 2
__proto__: Array(0)
type: "and"
__proto__: Object
__proto__: Object
```

</details>

**Comment by elephi** [comment]
Yes I got a similar issue.

When 
1. there is a field in the model is a List [] 
2. one of the records has a empty list

the Datastore.query which is using 'contains' will cause an error with different hint in different browser :

**Cannot read property 'indexOf' of null** in Chrome
**TypeError: value is null** in Firefox

**Comment by myfairshare** [comment]
Have the same issue, a real pain it is.

**Comment by elephi** [comment]
> Have the same issue, a real pain it is.

My temporary solution:
Add a dummy item to the list when initialize the record

**Comment by myfairshare** [comment]
> > Have the same issue, a real pain it is.
> 
> My temporary solution:
> Add a dummy item to the list when initialize the record

Will try that, thanks for the tip elephi!

**Comment by myfairshare** [comment]
> > Have the same issue, a real pain it is.
> 
> My temporary solution:
> Add a dummy item to the list when initialize the record

Thanks again for that, feeling a bit silly now.  Totally forgot to take into account my search was searching instances that might be null, all good now!!!

**Comment by DevTGhosh** [comment]
@chrisbonifacio why was this closed this is still an issue. I hope the official solution isn't putting dummy values in list on all items in a table to make contains work?

**Comment by nickarocho** [comment]
Hi @DevTGhosh thank you for raising your problem with us.

**TL;DR** - this is an issue with the older version of Amplify Datastore you are using (as seen in your Environment Info provided):
```
@aws-amplify/core: ^3.8.9 => 3.8.21 
@aws-amplify/datastore: ^2.9.14 => 2.9.15 
```

I was able to reproduce the problem _only after bumping down to this older version_, but with my app **using a more recent version of Amplify Core (4.X) and Amplify Datastore (3.X), this is not an issue** - the `contains` predicate within `DataStore.query` works as expected without failure.

For reference, I am using:
```
"@aws-amplify/core": "^4.0.2",
"@aws-amplify/datastore": "^3.0.2",
```

---

While upgrading your core library version to `4.X` and the respective Datastore version to `3.X` would fix the issue, I realize this might not be an option based on your current app setup and the corresponding potential breaking changes of a major version bump (though most of the breaking changes in Datastore 3.X are SDK related and most likely won't be a problem for you, keep an eye out for any **named imports**)

If upgrading library versions is not an option, I would suggest the intermediary solution of wrapping the line which you call `DataStore.query` and the associated `contains` predicate in a `try/catch` block so the rest of the code within the function can still run after this line fails. 

Something like:
```js
let projectData;

try {
  projectData = await DataStore.query(Projects, (c) =>
    c.freelancers("contains", auth.userInfo.tenantId).activeFlag("eq", true)
  );
} catch (error) {
  // handle/log the error
  console.log(error)
}
// now this line will still work properly even if the code in the "try" block fails
console.log({projectData})
```
should do the trick and at least gracefully handle any problems while letting the rest of your code run.

Hope this helps solve your problem and unblock you and your team. Please reach back out if you have any further questions which we can assist with!
